### PR TITLE
Default client scope handling

### PR DESCRIPTION
**Pull request by Age15990**
I missed some functionality for client scope handling and added it. Tests have been updated. If there is anything missing please let me know!

### CODE REVIEW

File: client.go [hunk 1]
```diff
@@ -496,6 +496,98 @@ func (client *gocloak) GetClient(token string, realm string, clientID string) (*
 	return &result, nil
 }
 
+// GetClientsDefaultScopes returns a list of the client's default scopes
+func (client *gocloak) GetClientsDefaultScopes(token string, realm string, clientID string) ([]ClientScope, error) {
```
Comment (SVilgelm) [inline client.go]: Please rewrite function to return pointers, not objects. Take a look at other functions, all functions return pointers now. For example: https://github.com/Nerzal/gocloak/blob/master/client.go#L555

```go
func (client *gocloak) GetClientsDefaultScopes(token string, realm string, clientID string) ([]*ClientScope, error) {
     var result []*ClientScope
```

File: client.go [hunk 2]
```diff
@@ -496,6 +496,98 @@ func (client *gocloak) GetClient(token string, realm string, clientID string) (*
...
+// RemoveDefaultScopeFromClient removes a client scope from the list of client's default scopes
+func (client *gocloak) RemoveDefaultScopeFromClient(token string, realm string, clientID string, scopeID string) error {
+	resp, err := client.getRequestWithBearerAuth(token).
+		Delete(client.getAdminRealmURL(realm, "clients", clientID, "default-client-scopes", scopeID))
+
+	return checkForError(resp, err)
+}
+
+// GetClientsOptionalScopes returns a list of the client's optional scopes
+func (client *gocloak) GetClientsOptionalScopes(token string, realm string, clientID string) ([]ClientScope, error) {
```
Comment (SVilgelm) [inline client.go]: ditto

File: client.go [hunk 3]
```diff
@@ -496,6 +496,98 @@ func (client *gocloak) GetClient(token string, realm string, clientID string) (*
...
+// RemoveOptionalScopeFromClient deletes a client scope from the list of client's optional scopes
+func (client *gocloak) RemoveOptionalScopeFromClient(token string, realm string, clientID string, scopeID string) error {
+	resp, err := client.getRequestWithBearerAuth(token).
+		Delete(client.getAdminRealmURL(realm, "clients", clientID, "optional-client-scopes", scopeID))
+
+	return checkForError(resp, err)
+}
+
+// GetDefaultOptionalClientScopes returns a list of the client's optional scopes
+func (client *gocloak) GetDefaultOptionalClientScopes(token string, realm string) ([]ClientScope, error) {
```
Comment (SVilgelm) [inline client.go]: ditto

File: client.go [hunk 4]
```diff
@@ -496,6 +496,98 @@ func (client *gocloak) GetClient(token string, realm string, clientID string) (*
...
+
+	if err := checkForError(resp, err); err != nil {
+		return nil, err
+	}
+
+	return result, nil
+}
+
+// GetDefaultDefaultClientScopes returns a list of the client's optional scopes
+func (client *gocloak) GetDefaultDefaultClientScopes(token string, realm string) ([]ClientScope, error) {
```
Comment (SVilgelm) [inline client.go]: ditto

File: client.go [hunk 5]
```diff
@@ -511,6 +603,21 @@ func (client *gocloak) GetClientScope(token string, realm string, scopeID string
 	return &result, nil
 }
 
+// GetClientScopes returns all client scopes
+func (client *gocloak) GetClientScopes(token string, realm string) ([]ClientScope, error) {
```
Comment (SVilgelm) [inline client.go]: ditto

File: client.go [hunk 6]
```diff
@@ -496,6 +496,98 @@ func (client *gocloak) GetClient(token string, realm string, clientID string) (*
...
+		Get(client.getAdminRealmURL(realm, "default-optional-client-scopes"))
+
+	if err := checkForError(resp, err); err != nil {
+		return nil, err
+	}
+
+	return result, nil
+}
+
+// GetDefaultDefaultClientScopes returns a list of the client's optional scopes
```
Comment (SVilgelm) [inline client.go]: probably it should be: `returns a list of the client's default-default scopes`?

File: client_test.go [hunk 7]
```diff
@@ -19,6 +19,7 @@ import (
 	"time"
 
 	"github.com/dgrijalva/jwt-go"
+	"github.com/stretchr/testify/assert"
```
Comment (SVilgelm) [inline client_test.go]: I'm ok with this import, but then we need to rewrite all tests to use testify instead of `FailIfErr` and etc functions.
> FailIfError -> assert.NoError
> FailIfNotErr -> assert.Error
> FailIf -> assert.Fail
> AssertEquals -> assert.Equals
> AssertNotEquals -> assert.NotEqual

I didn't know about testify library at a time I implemented the unit tests.
But this can be and should be done in a separate pull request.

File: client.go (hunk 6 above)
Comment (SVilgelm) [reply client.go]: I mean not `optional`

File: client_test.go (hunk 7 above)
Comment (Age15990) [reply client_test.go]: Hm I know, I wanted to use your assert methods, but unfortunately the deep assert that is happening in there tells me my 1 is not a 1 and my 0 is not a 0. I understand that replacing everything is unrelated and a fairly huge task, but if you want to, I can also do that. What do you suggest?

Comment (SVilgelm) [reply client_test.go]: If you have a time, you can do it :) Or we can let it be as it is and replace it later. Or you can use `assert.` function in the new unit tests.

Comment (Age15990) [reply client_test.go]: I thought it would not be that complicated to replace it, but unfortunately it is because we have to distinguish between critical functions (such as create and delete) that build tests or tidy up and others that just have to be checked. If critical changes fail, I think we have to use `assert.FailNow()` or similar... I have started implementing but I would for now prefer to leave as is, if that is okay for you. Updated the README anyways!

Comment (SVilgelm) [reply client_test.go]: Thank you!

**Comment by SVilgelm** [comment]
@Nerzal, it can be merged

**Comment by Nerzal** [comment]
Sorry, i was on vacation :D
